### PR TITLE
StakeDAO recursive collateral

### PR DESCRIPTION
**Pull request by tbrent**
Context: Intermediate `sqrt()` work got merged into the base branch already

Still have 2 custom tests I want to write

### CODE REVIEW

File: contracts/libraries/Fixed.sol [hunk 1]
```diff
@@ -675,4 +679,94 @@ function fullMul(uint256 x, uint256 y) pure returns (uint256 hi, uint256 lo) {
...
+///
+/// @dev See https://en.wikipedia.org/wiki/Methods_of_computing_square_roots#Babylonian_method.
+///
+/// Notes:
+/// - If x is not a perfect square, the result is rounded down.
+/// - Credits to OpenZeppelin for the explanations in comments below.
+///
+/// @param x The uint256 number for which to calculate the square root.
+/// @return result The result as a uint256.
+function sqrt256(uint256 x) pure returns (uint256 result) {
```
Comment (akshatmittal) [inline contracts/libraries/Fixed.sol:694]: I'm assuming this function is correct, not verifying the implementation.

File: contracts/plugins/assets/curve/CurveRecursiveCollateral.sol [hunk 2]
```diff
@@ -0,0 +1,192 @@
...
+        // {ref/tok} = {ref/lpToken} = {lpToken@t=0/lpToken} * {1} * 2{ref/lpToken@t=0}
+        return virtualPrice.mul(rTokenRate.sqrt()).mulu(2); // LP token worth twice as much
+    }
+
+    // === Internal ===
+
+    // Override this later to implement non-standard recursive pools
+    function _anyDepeggedInPool() internal view virtual override returns (bool) {
+        // Check reference token
+        try this.tokenPrice(0) returns (uint192 low, uint192 high) {
```
Comment (akshatmittal) [inline contracts/plugins/assets/curve/CurveRecursiveCollateral.sol:173]: Probably needs a comment saying that this only works where the first token is the paired token and the second token is the RToken.

File: contracts/plugins/assets/curve/stakedao/StakeDAORecursiveCollateral.sol [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        uint192 revenueHiding,
+        PTConfiguration memory ptConfig
+    ) CurveRecursiveCollateral(config, revenueHiding, ptConfig) {
+        IStakeDAOVault vault = IStakeDAOVault(address(config.erc20));
+        gauge = vault.liquidityGauge();
+        claimer = gauge.claimer();
+
+        uint256 rewardCount = gauge.reward_count();
+        for (uint256 i = 0; i < rewardCount; i++) {
+            rewardTokens.push(gauge.reward_tokens(i));
```
Comment (akshatmittal) [inline contracts/plugins/assets/curve/stakedao/StakeDAORecursiveCollateral.sol:58]: AFAIK, this list can change, should move this to the `claimRewards()` function. (eg: a pool with no SDT rewards can add SDT later)

File: contracts/plugins/assets/curve/stakedao/StakeDAORecursiveCollateral.sol [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+interface IStakeDAOGauge {
+    function claimer() external view returns (IStakeDAOClaimer);
+
+    function reward_count() external view returns (uint256);
+
+    function reward_tokens(uint256 index) external view returns (IERC20Metadata);
+}
+
+interface IStakeDAOClaimer {
```
Comment (akshatmittal) [inline contracts/plugins/assets/curve/stakedao/StakeDAORecursiveCollateral.sol:21]: Which claimer are you looking at? The one that StakeDAO currently uses on their website has two parameters, `guages` and `claimVeSDT`. [Contract](https://etherscan.io/address/0x539e65190a371ce73244a98dec42ba635cca512c).

**Review by akshatmittal — CHANGES_REQUESTED** [review]

File: contracts/plugins/assets/curve/stakedao/StakeDAORecursiveCollateral.sol (hunk 4 above)
Comment (tbrent) [reply contracts/plugins/assets/curve/stakedao/StakeDAORecursiveCollateral.sol:21]: the arg should be `false` yes?

**Review by akshatmittal — APPROVED** [review]

**Pull request merged**